src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1159,6 +1159,11 @@ object Types {
     private[this] var lastSymbol: Symbol = _
     private[this] var checkedPeriod = Nowhere
 
+    /** Provided for debugging, can be used to intercept assignments of uncommon denotations
+     */  
+    private def checkDenot(d: Denotation)(implicit ctx: Context) = {
```

Comment:
here https://github.com/lampepfl/dotty/commit/9a7a7b889c9ded58ceaefb90f6ee208c9d2ab534#diff-ca3386647f5386b6f70e99f1a9e188cdR1288
you can see how this can be done using what we already have in place. If you agree I would add this commit to PR.